File: Python3/1046. Last Stone Weight.py
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        # edge case
        if len(stones) == 1:
            return stones[0]
        
        stones = [-s for s in stones]
        heapq.heapify(stones)
        
        while len(stones) > 1:
            largest = heapq.heappop(stones)
            second = heapq.heappop(stones)
            
            # only push the remains if it's not 0
            if largest != second:
                heapq.heappush(stones, largest-second)
            
        return -stones[0] if stones else 0

File: Java/1046. Last Stone Weight.java
class Solution {
    public int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> heap = new PriorityQueue<>(Comparator.reverseOrder());
        for(int s: stones){
            heap.add(s);
        }
        
        while(heap.size() > 1){
            int largest = heap.poll();
            int second = heap.poll();
            
            if(largest != second){
                heap.add(largest-second);
            }
        }
        
        return heap.isEmpty() ? 0 : heap.poll();
    }
}